module-system/sov-modules-api/src/hooks.rs

```diff
@@ -31,14 +31,18 @@ pub trait ApplyBlobHooks {
     type BlobResult;
 
     /// Runs at the beginning of apply_blob, locks the sequencer bond.
+    /// If this hook returns Err, batch is not applied
+    /// TODO: Expectation about what is in working set, if it is "clean"
```

Comment:
This needs to be addresses

Reply:
I think the expectation is that a `WorkingSet` is dirty. If it's clean, we should pass a `StateCheckpoint` instead. 

cc @bkolad is that right?

Reply:
Where this expectations come from?  The idea is to create `WorkingSet` at the beginning of the `ApplyBatch` and pass it everywhere (why the functions we are passing it to have to expect anything about it?).

Reply:
> why the functions we are passing it to have to expect anything about it?

Because it receives mutable reference to working set. And in different hooks it gets different state of this working set.

For example, in `post_dispatch_tx_hook` it receives "dirty" working set, and it can be both commited or reverted, depending on transaction status, where in `post_apply_blob_hook` state is clean, which means was either commited or reveted.

Having no expectations about state makes it hard to use the hook. 
For example post dispatch hook for `Accounts` it looks like module assumes that state is clean and won't be reverted, so it increases account nonce on invalid transaction, when in reality it won't happen. I am working on updating tests to consider that case.

> The idea is to create WorkingSet at the beginning of the ApplyBatch and pass it everywhere

This approach is very complicated to implement without bugs, that's why it is good at least to have an assumptions

Reply:
Now, I think we could pass `StateCheckpoint` to the function and convert it to `WorkingSet` inside, so we add type-level information. But I am still not sure why the hooks need to be aware of why the `WorkinSet` was committed or reverted.

Reply:
> But I am still not sure why the hooks need to be aware of why the WorkinSet was committed or reverted.

I guess my point could be better formulated. It shouldn't be aware if it was commited or reverted, but it should expected if its changes are going to be commited or reverted. 

In my original comment, "dirty" state means that it will be reverted or commited after hook executed. Which introduces bug in nonce tracking.

Reply:
Yep, this makes sense. The more we  can use the type system the better IMO :)

Reply:
Yes, offloaded this to https://github.com/Sovereign-Labs/sovereign-sdk/issues/421